File: server/internal/api/prekeys.go
package api

import (
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/slimcord/slimcord-server/internal/auth"
	"github.com/slimcord/slimcord-server/internal/db"
)

type PrekeyHandler struct {
	authSvc *auth.AuthService
	db      *db.DB
}

func NewPrekeyHandler(authSvc *auth.AuthService, database *db.DB) *PrekeyHandler {
	return &PrekeyHandler{authSvc: authSvc, db: database}
}

// POST /api/v1/prekeys — Upload prekey bundle (auth required)
func (h *PrekeyHandler) HandleUpload(w http.ResponseWriter, r *http.Request) {
	userID, ok := r.Context().Value(auth.UserIDKey).(string)
	if !ok {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
		return
	}

	var req struct {
		IdentityKey          string   `json:"identity_key"`
		SignedPrekey         string   `json:"signed_prekey"`
		SignedPrekeySignature string  `json:"signed_prekey_signature"`
		OneTimePrekeys       []string `json:"one_time_prekeys"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid request body"})
		return
	}
	if req.IdentityKey == "" || req.SignedPrekey == "" || req.SignedPrekeySignature == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "identity_key, signed_prekey, and signed_prekey_signature are required"})
		return
	}

	// Serialize one_time_prekeys as JSON array.
	otpJSON, err := json.Marshal(req.OneTimePrekeys)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "failed to encode one_time_prekeys"})
		return
	}

	bundle := &db.PrekeyBundle{
		UserID:                userID,
		IdentityKey:           []byte(req.IdentityKey),
		SignedPrekey:          []byte(req.SignedPrekey),
		SignedPrekeySignature: []byte(req.SignedPrekeySignature),
		OneTimePrekeys:        otpJSON,
	}
	if err := h.db.SavePrekeyBundle(bundle); err != nil {
		slog.Error("save prekey bundle failed", "error", err)
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "failed to save prekey bundle"})
		return
	}

	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

// GET /api/v1/prekeys/{user_id} — Download a user's prekey bundle (auth required)
func (h *PrekeyHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
	targetUserID := r.PathValue("user_id")
	if targetUserID == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "user_id is required"})
		return
	}

	bundle, err := h.db.GetPrekeyBundle(targetUserID)
	if err != nil {
		slog.Error("get prekey bundle failed", "error", err)
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "internal server error"})
		return
	}
	if bundle == nil {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "prekey bundle not found"})
		return
	}

	// Consume one one-time prekey.
	var oneTimePrekeys []string
	otp, err := h.db.ConsumeOneTimePrekey(targetUserID)
	if err != nil {
		slog.Warn("consume one-time prekey failed", "error", err)
	}
	if otp != nil {
		oneTimePrekeys = []string{string(otp)}
	}

	resp := map[string]interface{}{
		"identity_key":           string(bundle.IdentityKey),
		"signed_prekey":          string(bundle.SignedPrekey),
		"signed_prekey_signature": string(bundle.SignedPrekeySignature),
		"one_time_prekeys":       oneTimePrekeys,
	}
	if oneTimePrekeys == nil {
		resp["one_time_prekeys"] = []string{}
	}

	writeJSON(w, http.StatusOK, resp)
}

// DELETE /api/v1/prekeys — Delete your prekey bundle (auth required)
func (h *PrekeyHandler) HandleDelete(w http.ResponseWriter, r *http.Request) {
	userID, ok := r.Context().Value(auth.UserIDKey).(string)
	if !ok {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
		return
	}

	if err := h.db.DeletePrekeyBundle(userID); err != nil {
		slog.Error("delete prekey bundle failed", "error", err)
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "failed to delete prekey bundle"})
		return
	}

	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

File: server-rs/src/api/prekeys.rs
use axum::{
    extract::{Path, State},
    Extension, Json,
};
use base64::Engine;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::api::AppState;
use crate::auth::UserId;
use crate::db;
use crate::error::AppError;

#[derive(Deserialize)]
pub struct UploadPrekeyRequest {
    pub identity_key: String,
    /// X25519 public DH key (base64). Required for X3DH's DH2 step;
    /// the Ed25519 `identity_key` above can't be used for raw DH.
    pub identity_dh_key: String,
    pub signed_prekey: String,
    pub signed_prekey_signature: String,
    #[serde(default)]
    pub one_time_prekeys: Vec<String>,
}

pub async fn upload(
    Extension(UserId(user_id)): Extension<UserId>,
    State(state): State<AppState>,
    Json(body): Json<UploadPrekeyRequest>,
) -> Result<Json<Value>, AppError> {
    let identity_key = base64::engine::general_purpose::STANDARD
        .decode(&body.identity_key)
        .map_err(|_| AppError::BadRequest("invalid base64 identity_key".into()))?;

    let identity_dh_key = base64::engine::general_purpose::STANDARD
        .decode(&body.identity_dh_key)
        .map_err(|_| AppError::BadRequest("invalid base64 identity_dh_key".into()))?;

    let signed_prekey = base64::engine::general_purpose::STANDARD
        .decode(&body.signed_prekey)
        .map_err(|_| AppError::BadRequest("invalid base64 signed_prekey".into()))?;

    let signed_prekey_signature = base64::engine::general_purpose::STANDARD
        .decode(&body.signed_prekey_signature)
        .map_err(|_| AppError::BadRequest("invalid base64 signed_prekey_signature".into()))?;

    // Store one-time prekeys as JSON array of base64 strings.
    let otpk_json = serde_json::to_vec(&body.one_time_prekeys)
        .map_err(|e| AppError::Internal(format!("serialize prekeys: {}", e)))?;

    let db = state.db.clone();
    let uid = user_id.clone();

    tokio::task::spawn_blocking(move || {
        db.with_conn(|conn| {
            let bundle = db::PrekeyBundle {
                id: db::new_id(),
                user_id: uid,
                identity_key,
                identity_dh_key,
                signed_prekey,
                signed_prekey_signature,
                one_time_prekeys: otpk_json,
                uploaded_at: db::now_str(),
            };
            db::save_prekey_bundle(conn, &bundle)
        })
    })
    .await
    .map_err(|e| AppError::Internal(format!("task join: {}", e)))?
    .map_err(|e| AppError::Internal(format!("db: {}", e)))?;

    Ok(Json(json!({ "ok": true })))
}

pub async fn get_bundle(
    Extension(UserId(_user_id)): Extension<UserId>,
    State(state): State<AppState>,
    Path(target_user_id): Path<String>,
) -> Result<Json<Value>, AppError> {
    let db = state.db.clone();
    let tuid = target_user_id.clone();

    let result = tokio::task::spawn_blocking(move || {
        db.with_conn(|conn| {
            let bundle = db::get_prekey_bundle(conn, &tuid)?
                .ok_or_else(|| rusqlite::Error::QueryReturnedNoRows)?;

            // Consume one OTP key if available.
            let one_time_prekey = db::consume_one_time_prekey(conn, &tuid)?;

            let identity_key_b64 =
                base64::engine::general_purpose::STANDARD.encode(&bundle.identity_key);
            let identity_dh_key_b64 =
                base64::engine::general_purpose::STANDARD.encode(&bundle.identity_dh_key);
            let signed_prekey_b64 =
                base64::engine::general_purpose::STANDARD.encode(&bundle.signed_prekey);
            let sig_b64 =
                base64::engine::general_purpose::STANDARD.encode(&bundle.signed_prekey_signature);
            // Wire format returns `one_time_prekeys` as an array
            // (length 0 or 1 — we consume at most one). Keeping the
            // field as an array means the client doesn't need a
            // singular/plural-aware parser.
            let otpk_array: Vec<String> = match one_time_prekey {
                Some(k) => vec![base64::engine::general_purpose::STANDARD.encode(&k)],
                None => vec![],
            };

            Ok(json!({
                "user_id": bundle.user_id,
                "identity_key": identity_key_b64,
                "identity_dh_key": identity_dh_key_b64,
                "signed_prekey": signed_prekey_b64,
                "signed_prekey_signature": sig_b64,
                "one_time_prekeys": otpk_array,
            }))
        })
    })
    .await
    .map_err(|e| AppError::Internal(format!("task join: {}", e)))?;

    match result {
        Ok(bundle) => Ok(Json(bundle)),
        Err(rusqlite::Error::QueryReturnedNoRows) => {
            Err(AppError::NotFound("prekey bundle not found".into()))
        }
        Err(e) => Err(AppError::Internal(format!("db: {}", e))),
    }
}

pub async fn delete_own(
    Extension(UserId(user_id)): Extension<UserId>,
    State(state): State<AppState>,
) -> Result<Json<Value>, AppError> {
    let db = state.db.clone();
    let uid = user_id.clone();

    tokio::task::spawn_blocking(move || {
        db.with_conn(|conn| db::delete_prekey_bundle(conn, &uid))
    })
    .await
    .map_err(|e| AppError::Internal(format!("task join: {}", e)))?
    .map_err(|e| AppError::Internal(format!("db: {}", e)))?;

    Ok(Json(json!({ "ok": true })))
}
